# ==== Question 1 =====

In [116]:


# import libraries
import pandas as pd
import numpy as np
import folium
import geopy
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
import requests # library to handle requests



Get the data from Wikipedia with pandas


In [117]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
list_df = pd.read_html(url)

The created list of Data Frames contain 3 data frames. The first one is the one we need. Let's assign it to a new data frame.



In [118]:
df = list_df[0]
print(df.shape)
# Display the first 10 rows
df.head()

(180, 3)


Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Let's rename the first column to PostalCode



In [119]:
df.rename(columns={"Postal Code":"PostalCode"},inplace=True)

# Get rid of rows with no Borough
to_drop = np.where(df['Borough']=='Not assigned')[0]
df.drop(to_drop,axis=0,inplace=True)
df.shape

(103, 3)

Let's check if any "Not assigned" values are still in Borough...

In [120]:
(df['Borough']=='Not assigned').sum()

0

Let's reset the index and check how the new DF looks like...

In [121]:
df = df.reset_index(drop=True)
df.head()

PostalCode           Borough                                Neighbourhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Let's check now for "Not assigned" values in Neighbourhood

In [122]:
(df['Neighbourhood']=="Not assigned").sum()

0

Now we combine rows that have the same PostalCode value

In [123]:
df = df.groupby(['PostalCode','Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.head()

PostalCode      Borough                           Neighbourhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [124]:
df.shape

(103, 3)

# ==== End of question 1 ====

# ==== Question 2 =====

 The geocoder package didn't work so we will load the csv data directly

In [125]:
coord_data = pd.read_csv('https://cocl.us/Geospatial_data')
coord_data.rename(columns={"Postal Code": "PostalCode" },inplace=True)

In [126]:
coord_data.head(10)

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476
5        M1J  43.744734 -79.239476
6        M1K  43.727929 -79.262029
7        M1L  43.711112 -79.284577
8        M1M  43.716316 -79.239476
9        M1N  43.692657 -79.264848

Now we merge both data frames into a single data frame....

In [127]:
new_df = df.merge(coord_data,on='PostalCode')
new_df.head(10)

PostalCode      Borough                                    Neighbourhood  \
0        M1B  Scarborough                                   Malvern, Rouge   
1        M1C  Scarborough           Rouge Hill, Port Union, Highland Creek   
2        M1E  Scarborough                Guildwood, Morningside, West Hill   
3        M1G  Scarborough                                           Woburn   
4        M1H  Scarborough                                        Cedarbrae   
5        M1J  Scarborough                              Scarborough Village   
6        M1K  Scarborough      Kennedy Park, Ionview, East Birchmount Park   
7        M1L  Scarborough                  Golden Mile, Clairlea, Oakridge   
8        M1M  Scarborough  Cliffside, Cliffcrest, Scarborough Village West   
9        M1N  Scarborough                      Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

# ==== End of question 2 ====

# ==== Question 3 ====

### First, we select the Boroughs that have the word "Toronto" and create a new data frame to work with...

In [128]:
neighborhoods = new_df[new_df['Borough'].str.contains("Toronto")].reset_index(drop=True)
print(neighborhoods.shape)
neighborhoods.head(10)

(39, 5)


PostalCode          Borough  \
0        M4E     East Toronto   
1        M4K     East Toronto   
2        M4L     East Toronto   
3        M4M     East Toronto   
4        M4N  Central Toronto   
5        M4P  Central Toronto   
6        M4R  Central Toronto   
7        M4S  Central Toronto   
8        M4T  Central Toronto   
9        M4V  Central Toronto   

                                       Neighbourhood   Latitude  Longitude  
0                                        The Beaches  43.676357 -79.293031  
1                       The Danforth West, Riverdale  43.679557 -79.352188  
2                     India Bazaar, The Beaches West  43.668999 -79.315572  
3                                    Studio District  43.659526 -79.340923  
4                                      Lawrence Park  43.728020 -79.388790  
5                                   Davisville North  43.712751 -79.390197  
6                  North Toronto West, Lawrence Park  43.715383 -79.405678  
7                                         Davisville  43.704324 -79.388790  
8                        Moore Park, Summerhill East  43.689574 -79.383160  
9  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049

There are 39 Neighbourhoods that we will be working with (note that most Neighbourhoods are a collection of many smaller neighbourhoods)

### Now we use geopy and get the coordinates for Toronto

In [129]:
from geopy.geocoders import Nominatim

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Use Folium to mark the 39 different neighbourhoods...

In [130]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Now we use the Foursquare IPA, Let's define Foursquare Credentials and Version

In [131]:
import foursquare_keys  # python script that contains my personal CLIENT IDs
CLIENT_ID = foursquare_keys.CLIENT_ID # your Foursquare ID
CLIENT_SECRET = foursquare_keys.CLIENT_SECRET # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

#### Let's create a function to get the nearby Venues to all the neighborhoods in Toronto

#### Note that we are using radius=10000 because we need enough distance to get several Venues per neighbourhood. 


In [132]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print("Venues found for",name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Apply the function to our Data Frame

In [133]:
toronto_venues = getNearbyVenues(names=neighborhoods['Neighbourhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

Venues found for The Beaches
Venues found for The Danforth West, Riverdale
Venues found for India Bazaar, The Beaches West
Venues found for Studio District
Venues found for Lawrence Park
Venues found for Davisville North
Venues found for North Toronto West, Lawrence Park
Venues found for Davisville
Venues found for Moore Park, Summerhill East
Venues found for Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Venues found for Rosedale
Venues found for St. James Town, Cabbagetown
Venues found for Church and Wellesley
Venues found for Regent Park, Harbourfront
Venues found for Garden District, Ryerson
Venues found for St. James Town
Venues found for Berczy Park
Venues found for Central Bay Street
Venues found for Richmond, Adelaide, King
Venues found for Harbourfront East, Union Station, Toronto Islands
Venues found for Toronto Dominion Centre, Design Exchange
Venues found for Commerce Court, Victoria Hotel
Venues found for Roselawn
Venues found for Forest Hill North & Wes

Let's see how many total Venues were found...

In [134]:
print(toronto_venues.shape)
toronto_venues.head()

(3176, 7)


Neighbourhood  Neighborhood Latitude  Neighborhood Longitude  \
0   The Beaches              43.676357              -79.293031   
1   The Beaches              43.676357              -79.293031   
2   The Beaches              43.676357              -79.293031   
3   The Beaches              43.676357              -79.293031   
4   The Beaches              43.676357              -79.293031   

               Venue  Venue Latitude  Venue Longitude  \
0  Glen Manor Ravine       43.676821       -79.293942   
1    Tori's Bakeshop       43.672114       -79.290331   
2     The Beech Tree       43.680493       -79.288846   
3  Beaches Bake Shop       43.680363       -79.289692   
4    The Fox Theatre       43.672801       -79.287272   

                  Venue Category  
0                          Trail  
1  Vegetarian / Vegan Restaurant  
2                      Gastropub  
3                         Bakery  
4            Indie Movie Theater

The total number of Venues is 3176 ... Now let's see how many Venues were returned for each neighbourhood

In [135]:
toronto_venues.groupby('Neighbourhood').count()[['Venue']]

Venue
Neighbourhood                                            
Berczy Park                                           100
Brockton, Parkdale Village, Exhibition Place          100
Business reply mail Processing Centre, South Ce...     50
CN Tower, King and Spadina, Railway Lands, Harb...     14
Central Bay Street                                    100
Christie                                              100
Church and Wellesley                                  100
Commerce Court, Victoria Hotel                        100
Davisville                                            100
Davisville North                                      100
Dufferin, Dovercourt Village                           64
First Canadian Place, Underground city                100
Forest Hill North & West, Forest Hill Road Park        49
Garden District, Ryerson                              100
Harbourfront East, Union Station, Toronto Islands     100
High Park, The Junction South                         100
India Bazaar, The Beaches West                         76
Kensington Market, Chinatown, Grange Park             100
Lawrence Park                                           8
Little Portugal, Trinity                              100
Moore Park, Summerhill East                            64
North Toronto West, Lawrence Park                      43
Parkdale, Roncesvalles                                100
Queen's Park, Ontario Provincial Government           100
Regent Park, Harbourfront                             100
Richmond, Adelaide, King                              100
Rosedale                                               22
Roselawn                                               26
Runnymede, Swansea                                     75
St. James Town                                        100
St. James Town, Cabbagetown                            33
Stn A PO Boxes                                        100
Studio District                                       100
Summerhill West, Rathnelly, South Hill, Forest ...     77
The Annex, North Midtown, Yorkville                   100
The Beaches                                            75
The Danforth West, Riverdale                          100
Toronto Dominion Centre, Design Exchange              100
University of Toronto, Harbord                        100

In [136]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 278 uniques categories.


Let's see which categories are the most common, and which are the least common...

In [137]:
pd.DataFrame({'Counts':pd.Series(toronto_venues['Venue Category'].value_counts())}).sort_values(by='Counts',ascending=False)[:10]

Counts
Coffee Shop             262
Café                    168
Restaurant               97
Italian Restaurant       91
Park                     87
Bakery                   71
Pizza Place              64
Japanese Restaurant      63
Sushi Restaurant         61
Hotel                    57

The most common Venues are coffee shops, Italian/Japanese rastaurants, Parks, Bakeries, Bars and hotels. 

In [138]:
pd.DataFrame({'Counts':pd.Series(toronto_venues['Venue Category'].value_counts())}).sort_values(by='Counts',ascending=True)[:10]

Counts
Residential Building (Apartment / Condo)       1
Rock Climbing Spot                             1
Tanning Salon                                  1
River                                          1
Sculpture Garden                               1
Taiwanese Restaurant                           1
Lingerie Store                                 1
Udon Restaurant                                1
Food                                           1
Dim Sum Restaurant                             1

The least common Venues are Exotic food restaurants  and other specific uncommon venues...

### Let's analyze each neighborhood with one hot enconding



In [139]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

print(toronto_onehot.shape)
toronto_onehot.head()

(3176, 279)


Neighbourhood  Accessories Store  Afghan Restaurant  Airport  \
0   The Beaches                  0                  0        0   
1   The Beaches                  0                  0        0   
2   The Beaches                  0                  0        0   
3   The Beaches                  0                  0        0   
4   The Beaches                  0                  0        0   

   American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
0                    0             0               0             0         0   
1                    0             0               0             0         0   
2                    0             0               0             0         0   
3                    0             0               0             0         0   
4                    0             0               0             0         0   

   Art Gallery  ...  Turkish Restaurant  Udon Restaurant  University  \
0            0  ...                   0                0           0   
1            0  ...                   0                0           0   
2            0  ...                   0                0           0   
3            0  ...                   0                0           0   
4            0  ...                   0                0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              1                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Whisky Bar  Wine Bar  Yoga Studio  Zoo  
0           0         0            0    0  
1           0         0            0    0  
2           0         0            0    0  
3           0         0            0    0  
4           0         0            0    0  

[5 rows x 279 columns]

### Now we group the data by Neighbourhoods and create the Data Frame we will use for the clustering function. 

In [140]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Berczy Park               0.00   
1        Brockton, Parkdale Village, Exhibition Place               0.01   
2   Business reply mail Processing Centre, South C...               0.00   
3   CN Tower, King and Spadina, Railway Lands, Har...               0.00   
4                                  Central Bay Street               0.00   
5                                            Christie               0.00   
6                                Church and Wellesley               0.00   
7                      Commerce Court, Victoria Hotel               0.00   
8                                          Davisville               0.00   
9                                    Davisville North               0.00   
10                       Dufferin, Dovercourt Village               0.00   
11             First Canadian Place, Underground city               0.00   
12    Forest Hill North & West, Forest Hill Road Park               0.00   
13                           Garden District, Ryerson               0.00   
14  Harbourfront East, Union Station, Toronto Islands               0.00   
15                      High Park, The Junction South               0.00   
16                     India Bazaar, The Beaches West               0.00   
17          Kensington Market, Chinatown, Grange Park               0.00   
18                                      Lawrence Park               0.00   
19                           Little Portugal, Trinity               0.00   
20                        Moore Park, Summerhill East               0.00   
21                  North Toronto West, Lawrence Park               0.00   
22                             Parkdale, Roncesvalles               0.00   
23        Queen's Park, Ontario Provincial Government               0.00   
24                          Regent Park, Harbourfront               0.00   
25                           Richmond, Adelaide, King               0.00   
26                                           Rosedale               0.00   
27                                           Roselawn               0.00   
28                                 Runnymede, Swansea               0.00   
29                                     St. James Town               0.00   
30                        St. James Town, Cabbagetown               0.00   
31                                     Stn A PO Boxes               0.00   
32                                    Studio District               0.00   
33  Summerhill West, Rathnelly, South Hill, Forest...               0.00   
34                The Annex, North Midtown, Yorkville               0.00   
35                                        The Beaches               0.00   
36                       The Danforth West, Riverdale               0.00   
37           Toronto Dominion Centre, Design Exchange               0.00   
38                     University of Toronto, Harbord               0.00   

    Afghan Restaurant   Airport  American Restaurant  Amphitheater  \
0                0.00  0.000000             0.020000          0.00   
1                0.00  0.000000             0.010000          0.00   
2                0.00  0.000000             0.020000          0.00   
3                0.00  0.071429             0.000000          0.00   
4                0.00  0.000000             0.010000          0.00   
5                0.00  0.000000             0.010000          0.00   
6                0.01  0.000000             0.010000          0.00   
7                0.00  0.000000             0.030000          0.00   
8                0.00  0.000000             0.000000          0.00   
9                0.00  0.000000             0.000000          0.00   
10               0.00  0.000000             0.000000          0.00   
11               0.00  0.000000             0.020000          0.00   
12               0.00  0.000000             0.000000          0.00   
13               0.

In [141]:
toronto_grouped.shape

(39, 279)

As expected, the DataFrame has 39 rows (neighbourhoods) and 278 venue columns (for each unique venue category)

### Let's see which are the 5 most common Venue categories in each Neighbourhood...

In [142]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                 venue  freq
0          Coffee Shop  0.11
1                 Café  0.06
2                Hotel  0.05
3           Restaurant  0.05
4  Japanese Restaurant  0.05


----Brockton, Parkdale Village, Exhibition Place----
                    venue  freq
0             Coffee Shop  0.07
1                    Café  0.07
2                     Bar  0.05
3                  Bakery  0.04
4  Furniture / Home Store  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0                Park  0.10
1         Coffee Shop  0.08
2             Brewery  0.06
3         Pizza Place  0.06
4  Italian Restaurant  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
             venue  freq
0      Coffee Shop  0.14
1             Café  0.14
2  Harbor / Marina  0.14
3     Dance Studio  0.07
4          Dog Run  0.07


----Central Bay Street----


### Now we create a new DataFrame that contains the 10 most common Venue categories for each Neighbourhood...

In [143]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place           Coffee Shop   
2  Business reply mail Processing Centre, South C...                  Park   
3  CN Tower, King and Spadina, Railway Lands, Har...           Coffee Shop   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
0                  Café   Japanese Restaurant                   Hotel   
1                  Café                   Bar  Furniture / Home Store   
2           Coffee Shop           Pizza Place                 Brewery   
3                  Café       Harbor / Marina                  Garden   
4      Sushi Restaurant                  Café                    Park   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant              Beer Bar                  Park   
1            Restaurant                Bakery    Tibetan Restaurant   
2    Italian Restaurant  Fast Food Restaurant      Sushi Restaurant   
3               Dog Run        Scenic Lookout                  Park   
4                 Hotel      Ramen Restaurant           Yoga Studio   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Cocktail Bar        Farmers Market     Seafood Restaurant  
1             Gift Shop    Italian Restaurant      Indian Restaurant  
2                Bakery       Harbor / Marina            Flea Market  
3                 Track      Sculpture Garden           Dance Studio  
4        Breakfast Spot       Bubble Tea Shop         Cosmetics Shop

### Finally, we use sklearn and run the K-means cluster function to the Neighborhoods...
We will use 5 clusters because a larger number results in many clusters with only 1 member, which may be an overfit model.


In [155]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check the distribution of cluster labels generated 
pd.DataFrame(pd.Series(kmeans.labels_[0:40]).value_counts()).rename(columns={0:'cluster size'})

cluster size
1            13
3             9
2             8
0             8
4             1

As seen from the distribution above, most clusters contain about 8-13 neighbourhoods, except for cluster label=4 where only there is only one neighbourhood.

### Now let's add the latitude, longitude, and Cluster label information to the DataFrame created above, so we can feed it to Folium and mark the clusters.

In [145]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() 

PostalCode          Borough                   Neighbourhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               3                   Pub           Coffee Shop   
1 -79.352188               1      Greek Restaurant           Coffee Shop   
2 -79.315572               1     Indian Restaurant           Coffee Shop   
3 -79.340923               0           Coffee Shop                   Bar   
4 -79.388790               4          College Quad           College Gym   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0           Pizza Place   Japanese Restaurant         Breakfast Spot   
1                  Café                   Pub                   Bank   
2         Grocery Store                  Park             Restaurant   
3   American Restaurant                 Diner  Vietnamese Restaurant   
4                  Park             Bookstore                   Café   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                 Beach                Bakery     Indian Restaurant   
1  Fast Food Restaurant    Italian Restaurant            Restaurant   
2                 Beach       Harbor / Marina                   Gym   
3                Bakery               Brewery    Italian Restaurant   
4  Gym / Fitness Center           Coffee Shop                 Trail   

  9th Most Common Venue       10th Most Common Venue  
0  Caribbean Restaurant                          Bar  
1      Ramen Restaurant                       Bakery  
2         Burrito Place         Fast Food Restaurant  
3      Sushi Restaurant                         Café  
4   Dumpling Restaurant  Eastern European Restaurant

### Let's display the map and mark the different clusters with different colors.

In [146]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
rainbow = ['red','blue','orange','green','black']


# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## We obtained a nice clustering of the neighbourhoods in the spatial dimension. This makes sense becuse we expect that all neighbourhoods downtown should be clustered together and also be distinct from neighbourhoods outside of downtown.

### Now, let's examine each cluster and determine the discriminating venue categories that distinguish each cluster. 



##  Cluster 0 (black points)

In [147]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
3                                Studio District               0   
25                University of Toronto, Harbord               0   
26     Kensington Market, Chinatown, Grange Park               0   
30                                      Christie               0   
31                  Dufferin, Dovercourt Village               0   
32                      Little Portugal, Trinity               0   
33  Brockton, Parkdale Village, Exhibition Place               0   
34                 High Park, The Junction South               0   

   1st Most Common Venue 2nd Most Common Venue          3rd Most Common Venue  \
3            Coffee Shop                   Bar            American Restaurant   
25                  Café                   Bar                         Bakery   
26                   Bar                  Café  Vegetarian / Vegan Restaurant   
30     Korean Restaurant                  Café                    Coffee Shop   
31                  Café           Coffee Shop                           Park   
32                  Café            Restaurant                            Bar   
33           Coffee Shop                  Café                            Bar   
34                  Café                   Bar                    Coffee Shop   

            4th Most Common Venue  5th Most Common Venue  \
3                           Diner  Vietnamese Restaurant   
25  Vegetarian / Vegan Restaurant     Mexican Restaurant   
26                    Coffee Shop            Yoga Studio   
30                  Grocery Store           Cocktail Bar   
31                            Bar       Sushi Restaurant   
32                         Bakery            Coffee Shop   
33         Furniture / Home Store             Restaurant   
34                Thai Restaurant      Convenience Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3                 Bakery               Brewery    Italian Restaurant   
25           Coffee Shop             Bookstore              Beer Bar   
26    Mexican Restaurant           Art Gallery         Grocery Store   
30    Mexican Restaurant  Ethiopian Restaurant                   Bar   
31    Italian Restaurant          Gourmet Shop               Brewery   
32    Italian Restaurant      Asian Restaurant           Pizza Place   
33                Bakery    Tibetan Restaurant             Gift Shop   
34                  Park      Sushi Restaurant                Bakery   

    9th Most Common Venue         10th Most Common Venue  
3        Sushi Restaurant                           Café  
25    Japanese Restaurant                            Pub  
26           Dessert Shop            Dumpling Restaurant  
30            Comedy Club                    Pizza Place  
31  Portuguese Restaurant                  Grocery Store  
32           Cocktail Bar  Vegetarian / Vegan Restaurant  
33     Italian Restaurant              Indian Restaurant  
34     Italian Restaurant                   Antique Shop

This cluster is mostly restaurants and cafes.

## Cluster 1 (red points)

In [157]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
1                        The Danforth West, Riverdale               1   
2                      India Bazaar, The Beaches West               1   
5                                    Davisville North               1   
6                   North Toronto West, Lawrence Park               1   
7                                          Davisville               1   
8                         Moore Park, Summerhill East               1   
9   Summerhill West, Rathnelly, South Hill, Forest...               1   
22                                           Roselawn               1   
23    Forest Hill North & West, Forest Hill Road Park               1   
24                The Annex, North Midtown, Yorkville               1   
35                             Parkdale, Roncesvalles               1   
36                                 Runnymede, Swansea               1   
38  Business reply mail Processing Centre, South C...               1   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
1       Greek Restaurant           Coffee Shop                  Café   
2      Indian Restaurant           Coffee Shop         Grocery Store   
5            Coffee Shop    Italian Restaurant            Restaurant   
6     Italian Restaurant    Mexican Restaurant                  Park   
7     Italian Restaurant           Coffee Shop      Sushi Restaurant   
8          Grocery Store           Coffee Shop    Italian Restaurant   
9            Coffee Shop      Sushi Restaurant    Italian Restaurant   
22    Italian Restaurant      Sushi Restaurant                  Café   
23                  Park                  Bank           Coffee Shop   
24    Italian Restaurant                  Café           Coffee Shop   
35                  Café           Pizza Place           Coffee Shop   
36           Coffee Shop                  Café                Bakery   
38                  Park           Coffee Shop           Pizza Place   

   4th Most Common Venue          5th Most Common Venue 6th Most Common Venue  \
1                    Pub                           Bank  Fast Food Restaurant   
2                   Park                     Restaurant                 Beach   
5                   Café           Fast Food Restaurant          Dessert Shop   
6            Coffee Shop                          Diner            Restaurant   
7             Restaurant                    Pizza Place  Fast Food Restaurant   
8                    Gym                Thai Restaurant            Restaurant   
9        Thai Restaurant                  Grocery Store                  Park   
22           Coffee Shop                       Pharmacy                  Bank   
23                  Café            Japanese Restaurant    Italian Restaurant   
24            Restaurant  Vegetarian / Vegan Restaurant                   Gym   
35        Breakfast Spot               Sushi Restaurant                   Bar   
36           Pizza Place             Italian Restaurant                   Pub   
38               Brewery             Italian Restaurant  Fast Food Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1     Italian Restaurant            Restaurant      Ramen Restaurant   
2        Harbor / Marina                   Gym         Burrito Place   
5       Sushi Restaurant                   Gym           Pizza Place   
6    Sporting Goods Shop                  Café          Skating Rink   
7      Indian Restaurant                  Café                   Gym   
8                   Park            Bagel Shop        Sandwich Place   
9             Restaurant           Pizza Place                  Bank   
22        Lingerie Store           Dry Cleaner        Clothing Store   
23                 Trail          Skating Rink      Sushi Restaurant   
24                Bakery                   Pub                Museum   
35         Grocery Store                Bakery                   Pub   
36       

This cluster is characterized by many restaurants, cafes, parks, and other commercial outlets like grocery store. Probably more residential than cluster 0. 

## Cluster 2 (blue points)

In [158]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
15                                     St. James Town               2   
16                                        Berczy Park               2   
18                           Richmond, Adelaide, King               2   
19  Harbourfront East, Union Station, Toronto Islands               2   
20           Toronto Dominion Centre, Design Exchange               2   
21                     Commerce Court, Victoria Hotel               2   
28                                     Stn A PO Boxes               2   
29             First Canadian Place, Underground city               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
15                  Café           Coffee Shop            Restaurant   
16           Coffee Shop                  Café   Japanese Restaurant   
18           Coffee Shop                  Café                 Hotel   
19           Coffee Shop                 Hotel                  Café   
20           Coffee Shop                 Hotel                  Café   
21           Coffee Shop                 Hotel            Restaurant   
28           Coffee Shop                  Café            Restaurant   
29           Coffee Shop                 Hotel                  Café   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
15   Japanese Restaurant                 Hotel              Beer Bar   
16                 Hotel            Restaurant              Beer Bar   
18               Theater            Restaurant                   Gym   
19            Restaurant                  Park   Japanese Restaurant   
20            Restaurant   Japanese Restaurant          Concert Hall   
21                  Café   Japanese Restaurant              Beer Bar   
28   Japanese Restaurant              Beer Bar                 Hotel   
29            Restaurant             Gastropub   Japanese Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
15    Seafood Restaurant                Bakery    Italian Restaurant   
16                  Park          Cocktail Bar        Farmers Market   
18          Concert Hall             Gastropub                Bakery   
19               Theater                   Gym        Scenic Lookout   
20               Theater              Beer Bar                  Park   
21             Gastropub   American Restaurant          Concert Hall   
28             Gastropub                  Park   American Restaurant   
29               Theater                  Park    Italian Restaurant   

           10th Most Common Venue  
15                      Gastropub  
16             Seafood Restaurant  
18            Japanese Restaurant  
19                        Brewery  
20            Monument / Landmark  
21  Vegetarian / Vegan Restaurant  
28             Seafood Restaurant  
29                       Beer Bar

This cluster is very distinct spatially in the map. There are many coffee shops and Hotels. This is probably very touristic area.

## Cluster 3 (orange points)

In [159]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
0                                         The Beaches               3   
10                                           Rosedale               3   
11                        St. James Town, Cabbagetown               3   
12                               Church and Wellesley               3   
13                          Regent Park, Harbourfront               3   
14                           Garden District, Ryerson               3   
17                                 Central Bay Street               3   
27  CN Tower, King and Spadina, Railway Lands, Har...               3   
37        Queen's Park, Ontario Provincial Government               3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                    Pub           Coffee Shop           Pizza Place   
10                  Park           Coffee Shop         Grocery Store   
11   Japanese Restaurant                 Diner             Gastropub   
12           Coffee Shop   Japanese Restaurant               Gay Bar   
13           Coffee Shop                  Café                  Park   
14           Coffee Shop             Gastropub    Italian Restaurant   
17           Coffee Shop      Sushi Restaurant                  Café   
27           Coffee Shop                  Café       Harbor / Marina   
37           Coffee Shop                  Park                  Café   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0    Japanese Restaurant        Breakfast Spot                 Beach   
10   Filipino Restaurant                Bistro        Sandwich Place   
11                  Café                  Park  Taiwanese Restaurant   
12                  Café      Sushi Restaurant                  Park   
13               Theater                   Pub            Restaurant   
14                 Hotel   Japanese Restaurant                 Diner   
17                  Park                 Hotel      Ramen Restaurant   
27                Garden               Dog Run        Scenic Lookout   
37      Sushi Restaurant      Ramen Restaurant           Pizza Place   

      7th Most Common Venue 8th Most Common Venue      9th Most Common Venue  \
0                    Bakery     Indian Restaurant       Caribbean Restaurant   
10                Juice Bar                  Bank        Japanese Restaurant   
11                   Garden    Italian Restaurant                 Restaurant   
12               Restaurant           Men's Store         Italian Restaurant   
13                   Bakery                 Diner             Breakfast Spot   
14  New American Restaurant             Bookstore  Middle Eastern Restaurant   
17              Yoga Studio        Breakfast Spot            Bubble Tea Shop   
27                     Park                 Track           Sculpture Garden   
37       Italian Restaurant               Gay Bar                 Restaurant   

   10th Most Common Venue  
0                     Bar  
10              BBQ Joint  
11          Jewelry Store  
12              Bookstore  
13      Indian Restaurant  
14                  Plaza  
17         Cosmetics Shop  
27           Dance Studio  
37            Yoga Studio

This cluster is also very commercial and has many restaurants.

## Cluster 4 (green points)

In [160]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  Lawrence Park               4          College Quad           College Gym   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4                  Park             Bookstore                  Café   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4  Gym / Fitness Center           Coffee Shop                 Trail   

  9th Most Common Venue       10th Most Common Venue  
4   Dumpling Restaurant  Eastern European Restaurant

This is a single neighbourhood. It is probably distinct from other neighbourhood because there is the College area and probably  a very residential area without many Venues nearby (foursquare found only 8 Venues with the 1000 meter radius).

### End of Notebook